pkg/clusteragent/autoscaling/cluster/controller.go

```diff
@@ -0,0 +1,430 @@
+
+	patchData := buildNodePoolPatch(np, mnp)
+	patchBytes, err := json.Marshal(patchData)
+	if err != nil {
+		return fmt.Errorf("error marshaling patch data: %s, err: %v", mnp.name, err)
+	}
+
+	// If NodePool is not considered a custom resource in the future, use StrategicMergePatchType
+	_, err = c.Client.Resource(nodePoolGVR).Patch(ctx, mnp.name, types.MergePatchType, patchBytes, metav1.PatchOptions{})
+	if err != nil {
+		return fmt.Errorf("unable to update NodePool: %s, err: %v", mnp.name, err)
+	}
+
+	return nil
+}
+
+func (c *Controller) deleteNodePool(ctx context.Context, name string) error {
+	log.Infof("Deleting NodePool: %s", name)
+
+	err := c.Client.Resource(nodePoolGVR).Delete(ctx, name, metav1.DeleteOptions{})
+	if err != nil {
+		return fmt.Errorf("Unable to delete NodePool: %s, err: %v", name, err)
+	}
+
+	return nil
+}
+
+type minNodePool struct {
+	name                     string            `json:"name"`
+	nodePoolHash             string            `json:"target_hash"`
+	recommendedInstanceTypes []string          `json:"recommended_instance_types"`
+	labels                   map[string]string `json:"labels"`
+	taints                   []corev1.Taint    `json:"taints"`
+}
+
+type minNodeClass struct {
+	Metadata minMetadata `json:"metadata"`
+}
+
+type minMetadata struct {
+	Name string `json:"name"`
+}
+
+// buildNodePoolSpec is used for creating new NodePools
+func buildNodePoolSpec(mnp minNodePool, nodeClassName string) karpenterv1.NodePoolSpec {
+
+	// Convert domain labels into requirements
+	reqs := []karpenterv1.NodeSelectorRequirementWithMinValues{}
+	for k, v := range mnp.labels {
+		reqs = append(reqs, karpenterv1.NodeSelectorRequirementWithMinValues{
+			NodeSelectorRequirement: corev1.NodeSelectorRequirement{
+				Key:      k,
+				Operator: corev1.NodeSelectorOpIn,
+				Values:   []string{v},
+			},
+		})
+	}
+
+	// Convert instance types into a requirement
+	reqs = append(reqs, karpenterv1.NodeSelectorRequirementWithMinValues{
+		NodeSelectorRequirement: corev1.NodeSelectorRequirement{
+			Key:      corev1.LabelInstanceTypeStable,
+			Operator: corev1.NodeSelectorOpIn,
+			Values:   mnp.recommendedInstanceTypes,
+		},
+	})
+
+	return karpenterv1.NodePoolSpec{
+		Template: karpenterv1.NodeClaimTemplate{
+			Spec: karpenterv1.NodeClaimTemplateSpec{
+				// Include taints
+				Taints:       mnp.taints,
+				Requirements: reqs,
+				NodeClassRef: &karpenterv1.NodeClassReference{
+					// Only support EC2NodeClass for now
+					Kind:  "EC2NodeClass",
+					Name:  nodeClassName,
+					Group: "karpenter.k8s.aws",
+				},
+			},
+		},
+	}
+}
+
+// buildNodePoolPatch is used to construct JSON patch
+func buildNodePoolPatch(np *karpenterv1.NodePool, mnp minNodePool) map[string]interface{} {
+
+	// Build requirements patch, only updating values for the instance types
+	newRequirements := []map[string]interface{}{}
+	for _, r := range np.Spec.Template.Spec.Requirements {
+		if r.Key == corev1.LabelInstanceTypeStable {
```

Comment:
does this mean if `LabelInstanceTypeStable` does not exist already in the requirements, we will not add it?